File: test/card-content.test.ts
import { describe, expect, it } from 'vitest';
import { extractRawCardText, isDegradedCardContent, parseRawCardWrapper } from '../src/bot/card-content';

// The real `json_card` returned by
//   GET /open-apis/im/v1/messages/:id?card_msg_content_type=raw_card_content
// for a 多维表格「发送消息卡片」automation card (captured live, slimmed).
const REAL_JSON_CARD = {
  body: {
    id: '_1',
    property: {
      elements: [
        {
          id: '_2',
          property: {
            elements: [
              { id: '_2_0', property: { userID: 'ou_ca62bad132e8a939051cc575a64c34e9' }, tag: 'at' },
              { id: '_2_1', property: { content: ' 请处理这条记录 ', textAlign: 'left' }, tag: 'plain_text' },
              {
                id: '_2_2',
                property: {
                  content: '查看杭州天气',
                  url: { url: 'https://my.feishu.cn/base/B9lIbXRuFaE7oOsP7q3cbzqqneg?table=tbl9QY31p4n9iS89&record=rec27E5vQq5cvL' },
                },
                tag: 'link',
              },
            ],
            originTag: 'markdown',
          },
          tag: 'markdown',
        },
        {
          id: '_3',
          property: {
            columns: [
              {
                id: '_4',
                property: {
                  elements: [
                    {
                      id: '_5',
                      property: {
                        actions: [{ action: { url: 'https://my.feishu.cn/base/B9lIbXRuFaE7oOsP7q3cbzqqneg?ccm_open_type=im_card_automation_button' }, type: 'open_url' }],
                        text: {
                          id: '_6',
                          property: { elements: [{ id: '_6_0', property: { content: '查看详情' }, tag: 'plain_text' }] },
                          tag: 'markdown',
                        },
                      },
                      tag: 'button',
                    },
                  ],
                },
                tag: 'column',
              },
            ],
          },
          tag: 'column_set',
        },
        { id: '_7', property: {}, tag: 'hr' },
        {
          id: '_8',
          property: {
            elements: [],
            i18nElements: {
              en_us: [
                { id: '_8_en_us_0', property: { content: 'From ' }, tag: 'plain_text' },
                { id: '_8_en_us_1', property: { content: 'Untitled base', url: { url: 'https://my.feishu.cn/base/B9lIbXRuFaE7oOsP7q3cbzqqneg?ccm_open_type=im_card_automation_link' } }, tag: 'link' },
              ],
              zh_cn: [
                { id: '_8_zh_cn_0', property: { content: '来自 ' }, tag: 'plain_text' },
                { id: '_8_zh_cn_1', property: { content: '未命名多维表格', url: { url: 'https://my.feishu.cn/base/B9lIbXRuFaE7oOsP7q3cbzqqneg?ccm_open_type=im_card_automation_link' } }, tag: 'link' },
              ],
            },
          },
          tag: 'markdown',
        },
      ],
    },
    tag: 'body',
  },
  header: {
    property: { title: { property: { content: '请关注记录' }, tag: 'plain_text' } },
    tag: 'card_header',
  },
  schema: '2.0',
};

describe('isDegradedCardContent', () => {
  it('flags the SDK fallback and the client-upgrade placeholder', () => {
    expect(isDegradedCardContent('[interactive card]')).toBe(true);
    expect(isDegradedCardContent('  ')).toBe(true);
    expect(isDegradedCardContent('请升级至最新版本客户端，以查看内容')).toBe(true);
  });
  it('leaves real content alone', () => {
    expect(isDegradedCardContent('请处理这条记录')).toBe(false);
  });
});

describe('parseRawCardWrapper', () => {
  it('unwraps the json_card envelope', () => {
    const body = JSON.stringify({ json_card: JSON.stringify({ schema: '2.0' }), json_attachment: {} });
    expect(parseRawCardWrapper(body)).toEqual({ schema: '2.0' });
  });
  it('returns the parsed body when not wrapped, undefined on garbage', () => {
    expect(parseRawCardWrapper('{"a":1}')).toEqual({ a: 1 });
    expect(parseRawCardWrapper('not json')).toBeUndefined();
  });
});

describe('extractRawCardText', () => {
  const text = extractRawCardText(REAL_JSON_CARD);

  it('leads with the card title', () => {
    expect(text.split('\n')[0]).toBe('请关注记录');
  });

  it('keeps the body text and the Base record link', () => {
    expect(text).toContain('请处理这条记录');
    expect(text).toContain('[查看杭州天气](https://my.feishu.cn/base/B9lIbXRuFaE7oOsP7q3cbzqqneg?table=tbl9QY31p4n9iS89&record=rec27E5vQq5cvL)');
  });

  it('keeps the button label and the footer source link', () => {
    expect(text).toContain('查看详情');
    expect(text).toContain('[未命名多维表格](https://my.feishu.cn/base/B9lIbXRuFaE7oOsP7q3cbzqqneg?ccm_open_type=im_card_automation_link)');
  });

  it('renders the footer in one locale only (no ×5 i18n duplication)', () => {
    expect(text).toContain('来自');
    expect(text).not.toContain('From ');
    expect(text).not.toContain('Untitled base');
  });

  it('drops the @ placeholder and empties (no bare open_id, no blank lines)', () => {
    expect(text).not.toContain('ou_ca62bad132e8a939051cc575a64c34e9');
    expect(text.split('\n').every((l) => l.trim().length > 0)).toBe(true);
  });

  it('yields nothing for an empty / unparseable card', () => {
    expect(extractRawCardText({})).toBe('');
    expect(extractRawCardText(null)).toBe('');
  });
});

File: internal/card/card_content_test.go
package card

import (
	"strings"
	"testing"
)

func TestIsDegradedCardContent(t *testing.T) {
	degraded := []string{"", "[interactive card]", "请升级至最新版本客户端，以查看内容", "please upgrade your client to view"}
	for _, s := range degraded {
		if !IsDegradedCardContent(s) {
			t.Errorf("should be degraded: %q", s)
		}
	}
	if IsDegradedCardContent("真实业务文本") {
		t.Fatal("real text should not be degraded")
	}
}

func TestParseRawCardWrapper_StringifiedJsonCard(t *testing.T) {
	// json_card 是字符串化 JSON。
	body := `{"json_card":"{\"property\":{\"content\":\"hi\"}}"}`
	v, ok := ParseRawCardWrapper(body)
	if !ok {
		t.Fatal("should parse")
	}
	m, _ := v.(map[string]any)
	if m["property"] == nil {
		t.Fatal("json_card should be unwrapped")
	}
}

func TestParseRawCardWrapper_ObjectDirect(t *testing.T) {
	body := `{"foo":"bar"}`
	v, ok := ParseRawCardWrapper(body)
	if !ok {
		t.Fatal("should parse non-wrapped object")
	}
	if v.(map[string]any)["foo"] != "bar" {
		t.Fatal("object passthrough wrong")
	}
}

func TestParseRawCardWrapper_BadJson(t *testing.T) {
	if _, ok := ParseRawCardWrapper("not json"); ok {
		t.Fatal("bad json should return false")
	}
}

func TestExtractRawCardText_ContentAndLink(t *testing.T) {
	card := map[string]any{
		"header": map[string]any{
			"property": map[string]any{"content": "卡片标题"},
		},
		"body": map[string]any{
			"property": map[string]any{
				"elements": []any{
					map[string]any{"property": map[string]any{"content": "请处理这条记录", "url": map[string]any{"url": "https://base.feishu.cn/x"}}},
				},
			},
		},
	}
	text := ExtractRawCardText(card)
	if !strings.Contains(text, "卡片标题") {
		t.Fatalf("missing title: %q", text)
	}
	if !strings.Contains(text, "请处理这条记录") {
		t.Fatalf("missing body content: %q", text)
	}
	if !strings.Contains(text, "[请处理这条记录](https://base.feishu.cn/x)") {
		t.Fatalf("link should pair content+url: %q", text)
	}
}

func TestExtractRawCardText_Dedup(t *testing.T) {
	card := map[string]any{
		"body": map[string]any{
			"property": map[string]any{
				"elements": []any{
					map[string]any{"property": map[string]any{"content": "重复"}},
					map[string]any{"property": map[string]any{"content": "重复"}},
				},
			},
		},
	}
	text := ExtractRawCardText(card)
	if strings.Count(text, "重复") != 1 {
		t.Fatalf("should dedup: %q", text)
	}
}

func TestExtractRawCardText_I18nOneLocale(t *testing.T) {
	card := map[string]any{
		"property": map[string]any{
			"i18nElements": map[string]any{
				"zh_cn": []any{map[string]any{"property": map[string]any{"content": "来自"}}},
				"en_us": []any{map[string]any{"property": map[string]any{"content": "from"}}},
			},
		},
	}
	text := ExtractRawCardText(card)
	if strings.Contains(text, "from") {
		t.Fatalf("en_us should not appear (zh_cn first): %q", text)
	}
	if !strings.Contains(text, "来自") {
		t.Fatalf("zh_cn content missing: %q", text)
	}
}
